metersink/output_odoo.py

```diff
@@ -129,157 +236,323 @@ def get_odoo_user_id(odoo):
+            mode="search_read",
+            filter_list=[
+                [
+                    ["order_id", "=", sale_orders["id"]],
+                    ["product_id", "=", 36],
+                ]
+            ],
+            # projection_dict=projection_dict
+        )
+        if line_record:
+            # todo update
+            # we are reading the last state of the line and calculate the time between this line and the update message.
+
+            if LOG.isEnabledFor(logging.DEBUG):
+                LOG.debug("%s", pformat(line_record))
+
+            time_calc = calculate_cloud_time(line_record["name"])
+
+            quantity = 3000
+            odoo_update(
+                odoo,
+                "sale.order.line",
+                line_record["id"],
+                {"product_uom_qty": quantity},
+            )
+        else:
+            # If there is no line already for the ressource, create it.
+
+            time_calc = calculate_cloud_time(
+                data_dict["traits"]["created_at"]
+            )
+
+            line_dict = {
+                "product_id": 36,
+                "name": display_name,
+                "product_uom_qty": time_calc,
+                "order_id": sale_orders["id"],
+            }
+            line_id = odoo_create(odoo, "sale.order.line", [line_dict])
+            LOG.debug("%s", line_id)
+
+
+def create_sale_order(odoo, customer, tag_list):
+    """creates a new SO for customer with tags"""
+    record_list = [
+                {
+                    "partner_id": customer['id'],
+                    "category_ids": tag_list
+                }
+            ]
+    new_id = odoo_create(odoo, "sale.order", record_list)
+    return new_id
+
+
+def create_sale_order_line(odoo, order_id, product_id, display_name, product_uom_qty):
+    """creates a new so line and returns its id"""
+    record_list = [
+        {
+            "order_id": order_id,
+            "product_id": product_id,
+            "display_name": display_name,
+            "product_uom_qty": product_uom_qty,
+        }
+    ]
+    new_id = odoo_create(odoo, "Sale.order.line", record_list)
+    return new_id
+
+
+def get_sale_order_id(odoo, tag_list):
     """
-    looks for already existing entries.
+    Returns the id of a Sale_order to further work on if the Customer is known.
+    Else it gives nothing back.
     """
-    # print(line)
-    if not filter_dict:
-        filter_dict = {}
+    # project_tag = f"project={project_id}"
+
+    [customer] = odoo_get_customer_from_project(odoo, tag_list)
+    if customer:
+        print(pformat(customer))
+        print(customer["sale_order_ids"])
+
+        if LOG.isEnabledFor(logging.DEBUG):
+            show_sale_order_fields(odoo)
+        filter_list = [
+            #"|",
+            ["customer_id" , "=", customer["id"]],
+            #["category_ids", "=", project_tag]
+        ]
+        projection_dict = get_projection_dict()
+        [sale_order_id] = odoo_get(
```

Comment:
Again the singleton list. Maybe there should be a function `odoo_get_item` that only returns the single element (and raises an exception whenever there are none or more than one)?

Reply:
since odoo is just a database with some ui and rpc-interface, there is always the possibility to have multiple or none result on a search. if we use limiting to one result it is still a list. I am not sure if a wrapper function to just take the one item out of the list is necessary. but maybe you are right. I will think about it. and put it on the list of things to do.

Reply:
Well, this line clearly states your expectation that this list has precisely one element. In any other case, it produces either `ValueError: too many values to unpack (expected 1)` or `ValueError: not enough values to unpack (expected 1, got 0)`. I think it would be nice to make this expectation a bit more explicit (and the code less cluttered) by using this wrapper function, which could then also produce better error messages. Even more so since it seems to be a very common case around here.